onnxruntime/core/framework/fallback_cpu_capability.cc

```diff
@@ -69,15 +69,13 @@ std::unordered_set<NodeIndex> GetCpuPreferredNodes(const onnxruntime::GraphViewe
     const Node* node = graph.GetNode(node_id);
 
     const KernelCreateInfo* kernel_info = kernel_lookup.LookUpKernel(*node);
-    // at least one registry has a target provider's kernel for this node
-    ORT_ENFORCE(kernel_info != nullptr);
     node_to_kernel.insert({node_id, kernel_info});
 
     // first, find all the direct consumer of cpu tensors.
     ORT_THROW_IF_ERROR(node->ForEachWithIndex(
         node->OutputDefs(),
         [&](const NodeArg& node_arg, size_t out_index) {
-          if (kernel_info->kernel_def->IsOutputOnCpu(out_index)) {
+          if (kernel_info == nullptr || kernel_info->kernel_def->IsOutputOnCpu(out_index)) {
```

Comment:
@fdwr I'd like to have some eyes from the ORT team just to confirm, but I believe this change is correct. Before this change, this part of the code was assuming that `kernel_info != nullptr` (which means that the EP supports the operator). 

This is generally true for other EPs like CUDA, but DML still doesn't support some operators. For example, let's say we have this series of operators in a graph:

`Div (int64) -> Add (int64) -> Mul (int64)`

If `Div` is placed on the CPU because, for example, its input is a small tensor or a shape tensor, it will also place Add and Mul on the CPU to avoid useless CPU<->GPU copies. This works because CUDA has an int64 registration for `Div`, which means that `kernel_info != nullptr`.

But in DML, since `Div` doesn't have an int64 registration yet, we cannot add `Div` to the `tentative_nodes`, which means that both `Add` and `Mul` will be placed on the GPU.

But is it `kernel_info != nullptr` a necessary condition? I don't think so. When `kernel_info == nullptr`, it means that the operator doesn't have a GPU EP registration and therefore MUST be on the CPU, which is the default fallback EP. And therefore, all of its inputs and outputs MUST be on the CPU, which means that `kernel_info == nullptr` is equivalent to `kernel_info->kernel_def->IsOutputOnCpu(out_index)`.

Reply:
I'm not sure who, but maybe @edgchen1 or @hariharans29 can help or would know who, given they touched this function.

Reply:
> When kernel_info == nullptr, it means that the operator doesn't have a GPU EP registration and therefore MUST be on the CPU, which is the default fallback EP. And therefore, all of its inputs and outputs MUST be on the CPU, which means that kernel_info == nullptr is equivalent to kernel_info->kernel_def->IsOutputOnCpu(out_index).

I am not sure about this. kernel_info == nullptr means that this EP doesn't have a matching kernel registration. As this code is run during EP::GetCapability(), I don't think we can assume that it also means the node will be assigned to the CPU EP. There may potentially be another EP that the node gets assigned to.

Reply:
Pat, I'm deferring to ORT folks for this part. The rest (DML EP parts) look fine to me.

Reply:
> > When kernel_info == nullptr, it means that the operator doesn't have a GPU EP registration and therefore MUST be on the CPU, which is the default fallback EP. And therefore, all of its inputs and outputs MUST be on the CPU, which means that kernel_info == nullptr is equivalent to kernel_info->kernel_def->IsOutputOnCpu(out_index).
> 
> I am not sure about this. kernel_info == nullptr means that this EP doesn't have a matching kernel registration. As this code is run during EP::GetCapability(), I don't think we can assume that it also means the node will be assigned to the CPU EP. There may potentially be another EP that the node gets assigned to.

@edgchen1 I see, thanks for the reply! Do you have a better suggestion to achieve this optimization then? This is a non-negligeable optimization in the models that we track and will get rid of many CPU->GPU copies. As a last resort, could we add an edge case for the DML EP? AFAIK, the DML EP is already not really compatible with other EPs, so at least for the DML EP, `kernel_info == nullptr` is equivalent to a CPU fallback. We could also copy the contents of the function in our own EP and make those changes, but then it could easily get out of sync.

Reply:
Maybe there could be an additional parameter to opt-in to this "no matching kernel implies CPU fallback" behavior.
@skottmckay what do you think?

Reply:
@edgchen1 @skottmckay I added a `requires_matching_kernel` argument that defaults to true, so the existing behavior will stay the same for all EPs.

Reply:
thanks for adding that. it would be good to state the assumptions made when `requires_matching_kernel == true` (like you've described above) in a comment in the code.

Reply:
I am thinking the logic have some problem. Take below two cases as example (RedcueProd has CPU kernel only):
- ...->Shape->Gather->ReduceProd->Concat->Reshape->...
- ...->Gather->ReduceProd->Concat->...

For case 1, it's OK we put Gather, ReduceProd and Concat to CPU, because they consumes output of Shape, which is small. But for case 2, suppose the Gather is consuming a big tensor, the ideal way is to keep ReduceProd in CPU only and all following nodes such as Concat should be in target EP. I.E, the expected graph should be ...->Gather->MemcpyToHost->ReduceProd->MemcpyFromHost->Concat->.... But after the change in this PR, since you put output of ReduceProd to cpu_output_args, it will also force Concat and may also some other following nodes to CPU, the final graph is ...->Gather->MemcpyToHost->ReduceProd->Concat->...->MemcpyFromHost->.... It's moving more nodes to CPU and will harm the perf.

From the code level, we use "if (kernel_info->kernel_def->IsOutputOnCpu(out_index))" to initialize cpu_output_args to make sure all of them added to it is small CPU tensors so that following subgraph is computed on small graph. But after the change, you also add output of not-kernel-found nodes to it, these outputs are on CPU, but may not be small, which will make following subgraph is computed on large tensors on CPU. Once this subgraph have more nodes, the perf is worse.

Reply:
I see. Do you have a model where the performance is affected by this? It seems that in this case, the solution would be to actually implement the missing operator for the DML EP, and therefore have the exact same placement as other EPs which also use this heuristic. From all the models I've analyzed, this change helps a lot by placing some `int64` ops on the CPU instead of adding a ton of unnecessary CPU<->GPU copies and it was needed to bring the device placement logic closer to other EPs.

I'd like to see examples where this logic makes perf worse so that I can investigate.

Reply:
I have same issue in CUDA as you so I have PR https://github.com/microsoft/onnxruntime/pull/13590 to fix. I don't have an example for the worse perf case, but logically it's possible. You can check my fix way of the issue, basically you cannot add any extra thing to cpu_output_args at the beginning when initialization, but can special handle it during travelling, so we can guarantee that we always start travelling from a small tensor.